docs/protocol_upgrade.md

```diff
@@ -0,0 +1,150 @@
+
+To determine whether a protocol feature is enabled, we do the following:
+- We maintain a `ProtocolFeature` enum where each variant corresponds to some protocol feature. For nightly protocol features,
+the variant is gated by the corresponding rust compile-time feature. For example, the evm feature is introduced as follows
+```rust
+pub enum ProtocolFeatures {
+  #[feature("protocol_feature_evm")],
+  EVM
+}
+```
+- We implement a function `protocol_version` to return, for each variant, the corresponding protocol version in which the
+feature is enabled. For example,
+```rust
+impl ProtocolFeature {
+    pub const fn protocol_version(self) -> ProtocolVersion {
+        match self {
+            // Stable features
+            // ...
+
+            // Nightly features
+            #[cfg(feature = "protocol_feature_evm")]
+            ProtocolFeature::EVM => 100,
+        }
+    }
+}  
+```
+- When we need to decide whether to use the new feature based on the protocol version of the current network,
+  we can simply compare it to the protocol version of the feature. For example,
+```rust
+#[cfg(feature = “protocol_version_evm”)]
+if current_protocol_version >= ProtocolFeature::EVM.protocol_version() {
+   do_evm_change();
+} else {
+   do_old_stuff();
+}
+#[cfg(not(feature = “protocol_version_evm”))] {
+   do_old_stuff();
+}
+```
+To make this simpler, we write a macro `checked_feature` that conveniently abstracts this logic:
+```rust
+#[macro_export]
+macro_rules! checked_feature {
+    ($feature_name:tt, $feature:ident, $current_protocol_version:expr) => {{
+        #[cfg(feature = $feature_name)]
+        let is_feature_enabled = $crate::version::ProtocolFeature::$feature.protocol_version()
+            <= $current_protocol_version;
+        #[cfg(not(feature = $feature_name))]
+        let is_feature_enabled = {
+            // Workaround unused variable warning
+            let _ = $current_protocol_version;
+
+            false
+        };
+        is_feature_enabled
+    }};
+
+    ($feature_name:tt, $feature:ident, $current_protocol_version:expr, $feature_block:block) => {{
+        checked_feature!($feature_name, $feature, $current_protocol_version, $feature_block, {})
+    }};
+
+    ($feature_name:tt, $feature:ident, $current_protocol_version:expr, $feature_block:block, $non_feature_block:block) => {{
+        #[cfg(feature = $feature_name)]
+        {
+            if checked_feature!($feature_name, $feature, $current_protocol_version) {
+                $feature_block
+            } else {
+                $non_feature_block
+            }
+        }
+        // Workaround unused variable warning
+        #[cfg(not(feature = $feature_name))]
+        {
+            let _ = $current_protocol_version;
+            $non_feature_block
+        }
+    }};
+}
+```
+With this macro, the previous example can be simplified into
+```rust
+checked_feature!(“protocol_feature_evm”, EVM, {do_evm_stuff();}, {do_old_stuff();});
+```
```

Comment:
To my taste, this is a bit too fine-grained and a bit too tied to specific API. I feel that this will get slightly wrong relatively quickly. 

I suggest trimming this section down to painting a big picture (we use compile-time features) and leaving *pointers* to more fine grained documentation ("see `ProtocolFeature` type and `checked_feature` enum for how to check features in code"). 

The specific API docs can then be moved to a doc string in an `.rs` file.